tests/api2/test_audit_basic.py

```diff
@@ -35,147 +84,156 @@ def initialize_for_smb_tests(request):
+        assert ds_config['org.freenas:quota_critical'] == AUDIT_CONFIG.defaults['quota_fill_critical']
+
+    def test_audit_config_updates(self):
+        """
+        This test validates that setting values has expected results.
+        """
+        new_config = call('audit.update', {'retention': 10})
+        assert new_config['retention'] == 10
+
+        # quota are in units of GiB
+        new_config = call('audit.update', {'quota': 1})
+        assert new_config['quota'] == 1
+        audit_dataset = call('audit.get_audit_dataset')
+
+        # ZFS value is in units of bytes.  Convert to GiB for comparison.
+        assert get_zfs('zfs', 'refquota', audit_dataset) // GiB == new_config['quota']
+
+        # Confirm ZFS and audit config are in sync
+        assert new_config['space']['available'] == get_zfs('space', 'available', audit_dataset)
+        assert new_config['space']['used_by_dataset'] == get_zfs('space', 'used', audit_dataset)
+
+        # Check that we're actually setting the quota by evaluating available space
+        # Change the the quota to something more interesting
+        new_config = call('audit.update', {'quota': 2})
+        assert new_config['quota'] == 2
+
+        audit_dataset = call('audit.get_audit_dataset')
+        assert get_zfs('zfs', 'refquota', audit_dataset) == 2*GiB  # noqa (allow 2*GiB)
+
+        used_in_dataset = get_zfs('space', 'used_by_dataset', audit_dataset)
+        assert abs(new_config['space']['available'] - 2*GiB) == used_in_dataset  # noqa (allow 2*GiB)
+
+        new_config = call('audit.update', {'reservation': 1})
+        assert new_config['reservation'] == 1
+        assert new_config['space']['used_by_reservation'] != 0
+
+        new_config = call('audit.update', {
+            'quota_fill_warning': 70,
+            'quota_fill_critical': 80
+        })
+
+        assert new_config['quota_fill_warning'] == 70
+        assert new_config['quota_fill_critical'] == 80
+
+        # Test disable reservation
+        new_config = call('audit.update', {'reservation': 0})
+        assert new_config['reservation'] == 0
+
+        # Test disable quota
+        new_config = call('audit.update', {'quota': 0})
+        assert new_config['quota'] == 0
+
+
+class TestAuditOps:
+    def test_audit_query(self, initialize_for_smb_tests):
+        share = initialize_for_smb_tests['share']
+        with smb_connection(
+            share=share['name'],
+            username=SMBUSER,
+            password=PASSWD,
+        ) as c:
+            fd = c.create_file('testfile.txt', 'w')
+            for i in range(0, 3):
+                c.write(fd, b'foo')
+                c.read(fd, 0, 3)
+            c.close(fd, True)
+
+        sleep(10)
+        ops = call('audit.query', {
+            'services': ['SMB'],
+            'query-filters': [['username', '=', SMBUSER]],
+            'query-options': {'count': True}
+        })
+        assert ops > 0
```

Comment:
Hmmm.
If we ran this section of code before the `with smb_connection`, would we get `ops` == 0?

If not, then we're not testing anything.  Would it make sense to run before and after and ensure the count increased?

Reply:
This is not a new test from me, but the test creates a temporary user, then calls `audit.query` _filtering for that user_.
So if the query and filters are working we should get non-zero only after creating the user.
Now, in general it's also good to do negative tests, so I'll add the same query before the user add and confirm `0`.
Thanks!